[webkit.RefCntblBaseVirtualDtor] Make ThreadSafeRefCounted not generate warnings

rniwa opened the pull request:
This PR makes WebKit's RefCntblBaseVirtualDtor checker not generate a warning for ThreadSafeRefCounted when the destruction thread is a specific thread.

Prior to this PR, we only allowed CRTP classes without a virtual destructor if its deref function had an explicit cast to the derived type, skipping any lambda declarations which aren't invoked. This ends up generating a warning for ThreadSafeRefCounted when a specific thread is used to destruct the object because there is no inline body / definition for ensureOnMainThread and ensureOnMainRunLoop and DerefFuncDeleteExprVisitor concludes that there is no explicit delete of the derived type.

This PR relaxes the condition DerefFuncDeleteExprVisitor checks by allowing a delete expression to appear within a lambda declaration if it's an argument to an "opaque" function; i.e. a function without definition / body.

haoNoQ commented on clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:
```diff
@@ -67,6 +68,15 @@ class DerefFuncDeleteExprVisitor
+        auto *Arg = CE->getArg(i);
+        VisitLambdaBody = true;
+        auto Restore = llvm::make_scope_exit([&] { VisitLambdaBody = false; });
+        if (VisitChildren(Arg))
```
You can probably avoid statefulness here if you directly pattern-match the argument as a lambda (with the help of a few manual `dyn_cast`s) and then invoke `Visit()` directly on that lambda's body. It'd also make `VisitLambdaExpr()` unnecessary.

It's usually not a good idea to search for a nested object at arbitrary nesting depths. When you do that, you lose your ability to confirm the exact relationship between the current object and the nested object (in our case, "the lambda is passed to the opaque call as an argument" as opposed to "the lambda is lexically nested into an argument in any shape or form"). The only time you actually want to ignore that relationship is in the very top-level visitor that simply looks for all matching code patterns in the entire translation unit no matter where they're written in the translation unit. It typically doesn't make a lot of sense in the middle of the pattern.

rniwa replied:
Okay, we can do that. Updated PR accordingly.

haoNoQ reviewed (commented):
Aha makes sense!

Looks like you're putting no restrictions on what the opaque function is. This may cause some false negatives but it's probably ultimately ok, but it might be a good idea to confirm.

rniwa commented:
> Looks like you're putting no restrictions on what the opaque function is. This may cause some false negatives but it's probably ultimately ok, but it might be a good idea to confirm.

Yes, we're not putting any requirement for the functions for now.

rniwa committed:
This PR makes WebKit's RefCntblBaseVirtualDtor checker not generate a warning for
ThreadSafeRefCounted when the destruction thread is a specific thread.

Prior to this PR, we only allowed CRTP classes without a virtual destructor
if its deref function had an explicit cast to the derived type, skipping any
lambda declarations which aren't invoked. This ends up generating a warning for
ThreadSafeRefCounted when a specific thread is used to destruct the object
because there is no inline body / definition for ensureOnMainThread and
ensureOnMainRunLoop and DerefFuncDeleteExprVisitor concludes that there is no
explicit delete of the derived type.

This PR relaxes the condition DerefFuncDeleteExprVisitor checks by allowing
a delete expression to appear within a lambda declaration if it's an argument
to ensureOnMainThread or ensureOnMainRunLoop.

rniwa commented:
> > Looks like you're putting no restrictions on what the opaque function is. This may cause some false negatives but it's probably ultimately ok, but it might be a good idea to confirm.
> 
> Yes, we're not putting any requirement for the functions for now.

Actually, why not. We can just require the function names to be ensureOnMainThread or ensureOnMainRunLoop for now. We can relax the requirement further if we find more cases.

haoNoQ commented on clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:
```diff
@@ -67,6 +68,48 @@ class DerefFuncDeleteExprVisitor
+
+  bool FindLabmda(const Expr *E) {
+    while (E) {
+      if (auto *TempE = dyn_cast<MaterializeTemporaryExpr>(E)) {
```
Aha great!

`Expr->IgnoreParenCasts()` skips all of these automatically, recursively too, except `CXXBindTemporaryExpr`. But IIRC `CXXBindTemporaryExpr` doesn't need to be skipped recursively anyway, it's a one-time thing that always appears immediately before the constructor.

rniwa replied:
Ah, that's so nice!

rniwa replied:
Simplified this code quite a bit using `IgnoreParenCasts`.

haoNoQ reviewed (approved):
Aha great LGTM!
